homeassistant/util/__init__.py

```diff
@@ -309,3 +309,53 @@ def wrapper(*args, **kwargs):
                 throttle[0].release()
 
         return wrapper
+
+
+class RetryOnError(object):
+    """A class for retrying a failed task a certain amount of tries.
+
+    This method decorator makes a method retrying on errors. If there was an
+    uncaught exception, it schedules another try to execute the task after a
+    retry delay. It does this up to the maximum number of retries.
+
+    It can be used for all probable "self-healing" problems like network
+    outages. The task will be rescheduled using HAs scheduling mechanism.
+
+    It takes a Hass instance, a maximum number of retries and a retry delay
+    in seconds as arguments.
+    """
+
+    def __init__(self, hass, retry_limit=0, retry_delay=20):
+        """Initialize the decorator."""
+        self.hass = hass
+        self.retry_limit = retry_limit
+        self.retry_delay = timedelta(seconds=retry_delay)
+
+    def __call__(self, method):
+        """Decorate the target method."""
+        from homeassistant.helpers.event import track_point_in_utc_time
+
+        @wraps(method)
+        def wrapper(*args, **kwargs):
+            """Wrapped method."""
+            def scheduled(retry=0, event=None):
+                """Call the target method.
+
+                It is called directly at the first time and then called
+                scheduled within the Hass mainloop.
+                """
+                # pylint: disable=broad-except
+                try:
+                    method(*args, **kwargs)
+                except Exception:
```

Comment:
I can change that. I thought it might be good to act on all exceptions that are not handled within the decorated method.